File: frontend/src/api/posture.ts
import { api } from './client';

export interface PostureCheckResult {
  rule_id: number;
  rule_name: string;
  category: string;
  severity: string;
  status: string;
  title: string;
  description: string;
  detection_id: number | null;
  context_data: Record<string, unknown>;
  triggered_at: string | null;
}

export interface RepoSummary {
  total: number;
  passing: number;
  warning: number;
  failing: number;
}

export interface RepoPosture {
  repo_name: string;
  org: string;
  visibility: string | null;
  default_branch: string | null;
  archived: boolean;
  fork: boolean;
  language: string | null;
  pushed_at: string | null;
  score: number;
  checks: PostureCheckResult[];
  detection_count: number;
}

export interface OrgPosture {
  org_login: string;
  score: number;
  two_factor_required: boolean | null;
  default_repo_permission: string | null;
  members_can_fork_private_repos: boolean | null;
  members_can_create_public_repos: boolean | null;
  ip_allow_list_enabled: boolean | null;
  checks: PostureCheckResult[];
  repos: RepoPosture[] | null;
  repo_summary: RepoSummary | null;
  detection_count: number;
}

export interface BreadcrumbItem {
  label: string;
  href: string | null;
}

export interface PostureResponse {
  level: string;
  score: number;
  orgs: OrgPosture[] | null;
  org: OrgPosture | null;
  repo: RepoPosture | null;
  breadcrumb: BreadcrumbItem[];
  last_sync_at: string | null;
  page: number;
  page_size: number;
  total: number;
  has_next: boolean;
}

export function getPosture(params?: { org?: string; repo?: string; search?: string; page?: number; page_size?: number }): Promise<PostureResponse> {
  return api.get<PostureResponse>('/posture', params as Record<string, string>);
}

File: backend/app/schemas/posture.py
"""Pydantic schemas for the security posture endpoint."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class PostureCheckResult(BaseModel):
    rule_id: int
    rule_name: str
    category: str
    severity: str
    status: str  # pass | fail | open | investigating | resolved | false_positive | unknown
    title: str
    description: str
    detection_id: int | None = None
    context_data: dict[str, Any] = {}
    triggered_at: datetime | None = None


class RepoSummary(BaseModel):
    total: int
    passing: int
    warning: int
    failing: int


class RepoPosture(BaseModel):
    repo_name: str
    org: str
    visibility: str | None = None
    default_branch: str | None = None
    archived: bool = False
    fork: bool = False
    language: str | None = None
    pushed_at: datetime | None = None
    score: float
    checks: list[PostureCheckResult]
    detection_count: int = 0


class OrgPosture(BaseModel):
    org_login: str
    score: float
    two_factor_required: bool | None = None
    default_repo_permission: str | None = None
    members_can_fork_private_repos: bool | None = None
    members_can_create_public_repos: bool | None = None
    ip_allow_list_enabled: bool | None = None
    checks: list[PostureCheckResult]
    repos: list[RepoPosture] | None = None
    repo_summary: RepoSummary | None = None
    detection_count: int = 0


class BreadcrumbItem(BaseModel):
    label: str
    href: str | None = None


class PostureResponse(BaseModel):
    level: str  # enterprise | org | repo
    score: float
    orgs: list[OrgPosture] | None = None
    org: OrgPosture | None = None
    repo: RepoPosture | None = None
    breadcrumb: list[BreadcrumbItem]
    last_sync_at: datetime | None = None
    page: int = 1
    page_size: int = 25
    total: int = 0
    has_next: bool = False
